File: Daily/141.linked-list-cycle.js
/*
 * @lc app=leetcode id=141 lang=javascript
 *
 * [141] Linked List Cycle
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {
    if (head === null) return false;
    let slow = head;
    let fast = head;

    while (fast !== null) {
        fast = fast.next;
        if (fast === null) break;
        fast = fast.next;

        slow = slow.next;

        if (slow === fast) return true;
    }

    return false;
};
// @lc code=end

File: prev-2024/141.linked-list-cycle.cpp
/*
 * @lc app=leetcode id=141 lang=cpp
 *
 * [141] Linked List Cycle
 */

struct ListNode {
  int val;
  ListNode* next;
  ListNode(int x) : val(x), next(nullptr) {}
};

// @lc code=start
class Solution {
 public:
  bool hasCycle(ListNode* head) {
    ListNode* fast = head;
    ListNode* slow = head;

    while (fast != nullptr && fast->next != nullptr) {
      fast = fast->next->next;
      slow = slow->next;

      if (slow == fast) {
        return true;
      }
    }

    return false;
  }
};
// @lc code=end
